src/common/row_set.h

```diff
@@ -110,19 +111,138 @@ class RowSetCollection {
       elem_of_each_node_.resize(right_node_id + 1, Elem(nullptr, nullptr, -1));
     }
 
-    elem_of_each_node_[left_node_id] = Elem(begin, split_pt, left_node_id);
-    elem_of_each_node_[right_node_id] = Elem(split_pt, e.end, right_node_id);
+    elem_of_each_node_[left_node_id] = Elem(begin, begin + n_left, left_node_id);
+    elem_of_each_node_[right_node_id] = Elem(begin + n_left, e.end, right_node_id);
     elem_of_each_node_[node_id] = Elem(nullptr, nullptr, -1);
   }
 
-  // stores the row indices in the set
+  // stores the row indexes in the set
   std::vector<size_t> row_indices_;
 
  private:
   // vector: node_id -> elements
   std::vector<Elem> elem_of_each_node_;
 };
 
+
+// The builder is required for samples partition to left and rights children for set of nodes
+// Responsible for:
+// 1) Effective memory allocation for intermediate results for multi-thread work
+// 2) Merging partial results produced by threads into original row set (row_set_collection_)
+// BlockSize is template to enable memory alignment easily with C++11 'alignas()' feature
+template<size_t BlockSize>
+class PartitionBuilder {
+ public:
+  template<typename Func>
+  void Init(const size_t n_tasks, size_t n_nodes, Func funcNTaks) {
```

Comment:
Any reason not to use a constructor? I strongly prefer constructors as it requires the object to be always in a usable state  and is much harder to misuse.

Reply:
Each time for new partition operation it reallocates buffers potentially, so we need to call Init() multiple times.